### PR TITLE
Extract security logic to PerftestSecurity class

### PR DESCRIPTION
**Pull request by javmorcas**

### CODE REVIEW

**Comment by javmorcas** [comment]
@softwarejimenez, do you mind taking a look to this code before we do the merge into the micro branch? This is the step prior to add the micro changes.

File: srcCpp/RTIDDSImpl.cxx [hunk 1]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
             "BaseProfileQos") != DDS_RETCODE_OK) {
-        fprintf(
-                stderr,
+        fprintf(stderr,
                 "Problem setting QoS Library \"%s::BaseProfileQos\" for participant_qos.\n",
                 _ProfileLibraryName);
     }
 
+    // Set initial peers and not use multicast
+    if ( _peer_host_count > 0 ) {
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
if (_peer_host_count > 0) {
```

File: srcCpp/RTIDDSImpl.cxx [hunk 2]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+        printf("Initial peers:\n");
+        for ( int i = 0; i < _peer_host_count; ++i) {
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
for (int i = 0; i < _peer_host_count; ++i) {
```

File: srcCpp/RTIDDSImpl.cxx [hunk 3]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+            printf("\t%s\n", _peer_host[i]);
+        }
+        qos.discovery.initial_peers.from_array(
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
 qos.discovery.initial_peers.from_array(
                (const char **)_peer_host,
                _peer_host_count);
```

File: srcCpp/RTIDDSImpl.cxx [hunk 4]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+
+    _factory = DDSDomainParticipantFactory::get_instance();
+
+    if (!ParseConfig(argc, argv)) {
+        return false;
+    }
+
+    // only if we run the latency test we need to wait
+    // for pongs after sending pings
+    _pongSemaphore = _LatencyTest ?
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
_pongSemaphore = _LatencyTest ? PerftestSemaphore_new() : NULL;
```

File: srcCpp/RTIDDSImpl.cxx [hunk 5]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+        NULL;
+
+    if (!ConfigureDomainParticipantQos(qos)) {
+        return false;
     }
 
-    // Creates the participant
     _participant = _factory->create_participant(
-        _DomainID, qos, listener,
+        (DDS_DomainId_t)_DomainID, qos, listener,
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
``` 
    _participant = _factory->create_participant(
            (DDS_DomainId_t)_DomainID, qos, listener,
            DDS_INCONSISTENT_TOPIC_STATUS |
            DDS_OFFERED_INCOMPATIBLE_QOS_STATUS |
            DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS);
```

File: srcCpp/RTIDDSImpl.cxx [hunk 6]
```diff
@@ -2529,7 +2153,7 @@ IMessagingReader *RTIDDSImpl<T>::CreateReader(
     if (_useUnbounded > 0) {
         char buf[10];
         sprintf(buf, "%lu", _useUnbounded);
-        DDSPropertyQosPolicyHelper::add_property(dr_qos.property,
+                DDSPropertyQosPolicyHelper::add_property(dr_qos.property,
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
char buf[10];
        sprintf(buf, "%lu", _useUnbounded);
        DDSPropertyQosPolicyHelper::add_property(
                dr_qos.property,
                "dds.data_reader.history.memory_manager.fast_pool.pool_buffer_max_size",
                buf, false);
```

File: srcCpp/Infrastructure_pro.cxx [hunk 7]
```diff
@@ -528,6 +526,195 @@ bool PerftestConfigureTransport(
...
+            file += "EncryptBoth";
+        } else if (security.dataEncrypted) {
+            file += "EncryptData";
+        } else if (security.subMessageEncrypted) {
+            file += "EncryptSubmessage";
+        }
+
+        file = file + ".xml";
+
+        fprintf(
```
Comment (softwarejimenez) [inline srcCpp/Infrastructure_pro.cxx]:
```
        fprintf(stdout,
                "\tUsing pre-built governance file: \n\t./%s\n", file.c_str());
```

File: srcCpp/PerftestSecurity.cxx [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    debugLevel(-1)
+{
+}
+
+/******************************************************************************/
+/* PUBLIC METHODS */
+
+std::map<std::string, unsigned int> PerftestSecurity::getSecurityCmdLineArgs()
+{
+
```
Comment (softwarejimenez) [inline srcCpp/PerftestSecurity.cxx]: Add a commit explaining it.
```
/*
 *    This map specifies if a parameter will be followed by another argument. 
 *    It is used to increment the index while parsing the parameter
 */

File: srcCpp/RTIDDSImpl.cxx [hunk 9]
```diff
@@ -2460,14 +2086,11 @@ IMessagingReader *RTIDDSImpl<T>::CreateReader(
...
-    {
-        if (_IsReliable)
-        {
+    if (strcmp(topic_name, perftest_cpp::_AnnouncementTopicName) != 0)  {
+        if (_IsReliable) {
             dr_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
         }
-        else
-        {
+        else {
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
} else {
```

File: srcCpp/perftest_publisher.cxx [hunk 10]
```diff
@@ -1904,8 +1904,15 @@ int perftest_cpp::Publisher()
...
         i++;
     }
-    reader_listener->print_summary_latency();
-    reader_listener->end_test = true;
+
+    if (_PubID == 0) {
+        reader_listener->print_summary_latency();
+        reader_listener->end_test = true;
+    } else {
+        fprintf(
```
Comment (softwarejimenez) [inline srcCpp/perftest_publisher.cxx]:
```
fprintf(stderr,
        "Latency results are only shown when -pidMultiPubTest = 0\n");
```

File: srcCpp/Infrastructure_pro.cxx [hunk 11]
```diff
@@ -528,6 +526,195 @@ bool PerftestConfigureTransport(
...
+
+        if (security.dataEncrypted && security.subMessageEncrypted) {
+            file += "EncryptBoth";
+        } else if (security.dataEncrypted) {
+            file += "EncryptData";
+        } else if (security.subMessageEncrypted) {
+            file += "EncryptSubmessage";
+        }
+
+        file = file + ".xml";
```
Comment (softwarejimenez) [inline srcCpp/Infrastructure_pro.cxx]: file += ".xml"

File: srcCpp/PerftestSecurity.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+public:
+
+    /**************************************************************************/
+    /* PUBLIC CLASS MEMBERS */
+
+    bool useSecurity;
+    bool signPackages;
+    bool dataEncrypted; // user data
+    bool subMessageEncrypted;   // submessage
```
Comment (softwarejimenez) [inline srcCpp/PerftestSecurity.h]:
```
bool subMessageEncrypted; // submessage
```

File: srcCpp/PerftestTransport.cxx [hunk 13]
```diff
@@ -4,12 +4,12 @@
  */
 
 #include "PerftestTransport.h"
+// #ifdef RTI_SECURE_PERFTEST
```
Comment (softwarejimenez) [inline srcCpp/PerftestTransport.cxx]: remove it

File: srcCpp/RTIDDSImpl.cxx [hunk 14]
```diff
@@ -702,6 +589,14 @@ bool RTIDDSImpl<T>::ParseConfig(int argc, char *argv[])
         return false;
     };
 
+  #ifdef RTI_SECURE_PERFTEST
+    if(!_security.parseSecurityOptions(argc, argv)) {
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
if (!_security.parseSecurityOptions(argc, argv)) {
```

File: srcCpp/RTIDDSImpl.cxx [hunk 15]
```diff
@@ -2040,58 +1636,88 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+    DDS_DomainParticipantFactoryQos factory_qos;
+    DomainListener *listener = new DomainListener();
+
+    _factory = DDSDomainParticipantFactory::get_instance();
+
+    if (!ParseConfig(argc, argv)) {
+        return false;
+    }
+
+    // only if we run the latency test we need to wait
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]: /*
     * Only if we run the latency test we need to wait
     * for pongs after sending pings
     */

File: srcCpp/Infrastructure_pro.cxx (hunk 11 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_pro.cxx]: Done.

File: srcCpp/Infrastructure_pro.cxx (hunk 7 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_pro.cxx]: Done.

File: srcCpp/Infrastructure_pro.cxx [hunk 16]
```diff
@@ -528,6 +526,195 @@ bool PerftestConfigureTransport(
...
+
+        fprintf(
+                stdout,
+                "\tUsing pre-built governance file: \n\t./%s\n",
+                file.c_str());
+        retcode = DDSPropertyQosPolicyHelper::add_property(
```
[the remark this answers is not in the document]
Comment (javmorcas) [reply srcCpp/Infrastructure_pro.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 14 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 9 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 5 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 4 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 15 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/perftest_publisher.cxx (hunk 10 above)
Comment (javmorcas) [reply srcCpp/perftest_publisher.cxx]: Done.

File: srcCpp/PerftestSecurity.cxx (hunk 8 above)
Comment (javmorcas) [reply srcCpp/PerftestSecurity.cxx]: Done.

File: srcCpp/PerftestSecurity.cxx [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+    << "\t-secureCertFile <file>        - Certificate file <optional>\n"
+    << "\t                                Default: \"./resource/secure/sub.pem\"\n"
+    << "\t-securePrivateKey <file>      - Private key file <optional>\n"
+    << "\t                                Default: \"./resource/secure/subkey.pem\"\n";
+    return oss.str();
+}
+
+bool PerftestSecurity::parseSecurityOptions(int argc, char *argv[])
+{
+    // We will only parse the properties related with transports here.
```
[the remark this answers is not in the document]
Comment (javmorcas) [reply srcCpp/PerftestSecurity.cxx]: Done

File: srcCpp/PerftestSecurity.cxx [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+    stringStream << "\tCertificate authority file: "
+                 << (certAuthorityFile.empty() ? "Not Specified" : certAuthorityFile)
+                 << "\n";
+
+    stringStream << "\tPlugin library: "
+                 << (library.empty() ? "Not Specified" : library)
+                 << "\n";
+
+    if( debugLevel != -1 ){
```
[the remark this answers is not in the document]
Comment (javmorcas) [reply srcCpp/PerftestSecurity.cxx]: Done

File: srcCpp/PerftestSecurity.h (hunk 12 above)
Comment (javmorcas) [reply srcCpp/PerftestSecurity.h]: Done.

File: srcCpp/PerftestTransport.cxx (hunk 13 above)
Comment (javmorcas) [reply srcCpp/PerftestTransport.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 1 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 2 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 3 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx [hunk 19]
```diff
@@ -2040,58 +1635,93 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
-        printf("Initial peers:\n");
-        for ( int i = 0; i< _peer_host_count; ++i) {
-            printf("\t%s\n", _peer_host[i]);
-        }
-        qos.discovery.initial_peers.from_array(
-            (const char **)_peer_host,
-            _peer_host_count);
-        qos.discovery.multicast_receive_addresses.length(0);
+    if (_AutoThrottle) {
+        DDSPropertyQosPolicyHelper::add_property(
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]: Why do not we check retcode here. 

```
        retcode = DDSPropertyQosPolicyHelper::add_property(
                qos.property,
                "dds.domain_participant.auto_throttle.enable",
                "true",
                false);
        if (retcode != DDS_RETCODE_OK) {
            printf("Failed to add property "
                    "dds.domain_participant.auto_throttle.enable\n");
            return false;
        }
```
And can you do it in the all the add_property calls?

File: srcCpp/RTIDDSImpl.cxx [hunk 20]
```diff
@@ -2040,58 +1635,93 @@ bool RTIDDSImpl<T>::Initialize(int argc, char *argv[])
...
+    _pongSemaphore = _LatencyTest ? PerftestSemaphore_new() : NULL;
+
+    if (!ConfigureDomainParticipantQos(qos)) {
+        return false;
     }
 
-    // Creates the participant
     _participant = _factory->create_participant(
-        _DomainID, qos, listener,
+        (DDS_DomainId_t) _DomainID,
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]:
```
    _participant = _factory->create_participant(
            (DDS_DomainId_t) _DomainID,
            qos,
            listener,
            DDS_INCONSISTENT_TOPIC_STATUS |
            DDS_OFFERED_INCOMPATIBLE_QOS_STATUS |
            DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS);
```

File: srcCpp/RTIDDSImpl.cxx [hunk 21]
```diff
@@ -2295,9 +1924,11 @@ IMessagingWriter *RTIDDSImpl<T>::CreateWriter(const char *topic_name)
     if (_useUnbounded > 0) {
         char buf[10];
         sprintf(buf, "%lu", _useUnbounded);
-        DDSPropertyQosPolicyHelper::add_property(dw_qos.property,
-               "dds.data_writer.history.memory_manager.fast_pool.pool_buffer_max_size",
-               buf, false);
+        DDSPropertyQosPolicyHelper::add_property(
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]: check retcode

File: srcCpp/RTIDDSImpl.cxx [hunk 22]
```diff
@@ -2529,9 +2157,12 @@ IMessagingReader *RTIDDSImpl<T>::CreateReader(
     if (_useUnbounded > 0) {
         char buf[10];
         sprintf(buf, "%lu", _useUnbounded);
-        DDSPropertyQosPolicyHelper::add_property(dr_qos.property,
-                "dds.data_reader.history.memory_manager.fast_pool.pool_buffer_max_size",
-                buf, false);
+        DDSPropertyQosPolicyHelper::add_property(
```
Comment (softwarejimenez) [inline srcCpp/RTIDDSImpl.cxx]: check retcode

File: srcCpp/RTIDDSImpl.cxx (hunk 19 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 20 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 21 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

File: srcCpp/RTIDDSImpl.cxx (hunk 22 above)
Comment (javmorcas) [reply srcCpp/RTIDDSImpl.cxx]: Done.

**Pull request closed without merging**